Key management updates in "Testing cryptography", "Testing data storage" for iOS and Android

vixentael opened the pull request:
I slightly refreshed crypto chapter:

* Updated "Protecting Keys in Memory" adding more information of keys clean up and key wrapping techniques. 
* Clarified IV usage, add note about popular mistake of copy-pasting example code with hardcoded IVs from OSS crypto libs.
* Mentioned BLAKE3.
* Added @veorq cryptocoding guidelines as reference for zeroing memory.

These changes were discussed with @sushi2k and @julepka in a separate doc.

---

Thank you for submitting a Pull Request to the Mobile Security Testing Guide. Please make sure that:

- [x] Your contribution is written in the 2nd person (e.g. you)
- [x] Your contribution is written in an active present form for as much as possible.
- [x] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [x] Your contribution has proper formatted markdown and/or code
- [x] Any references to website have been formatted as [TEXT](URL “NAME”)
- [x] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

vixentael commented:
markdown check is failing, but due to links in other places
<img width="953" alt="Screenshot 2021-03-26 at 15 44 22" src="https://user-images.githubusercontent.com/2877920/112640640-31410180-8e4a-11eb-85b4-7284863a91fd.png">

vixentael commented:
the Great BLAKE creator @veorq told me that BLAKE should be written as BLAKE not Blake. I obey.

cpholguera reviewed (changes requested):
hi @vixentael, thanks a lot for the nice PR. Please take a look at the change suggestions and the comments. There sre mainly 2 points where a little bit research would be great. It's very nice that you brought this up since the guide really needs to give clearer recommendations on this. I hope the reference links I've included help you come up with something. Could you please enhance the section with that? Thanks a lot in advance!

vixentael commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -160,10 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
-- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. Another approach is obfuscation: obfuscat ...
-- Make sure that keys live in memory for the shortest time possible. Ideally, keys should be read/allocated right before usage and clean up after usage or in case of error. Never store keys in constants.
-- Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap. See [Clean memory of secret data](https://github.com/veorq/cryptocoding#clean-memory-of-secret-data/ "The Cryptocoding Guidelines by @ve ...
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use [Android Keystore](https://developer.android.com/training/articles/keystore.html "Android keystore system")) or [Secure Enclave](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave "Storing Keys in the Secure ...
```
@cpholguera FYI: added official links here

vixentael commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -158,8 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
-- make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use Android Keystore) or Secure Enclave (e.g. use the Keychain and when you sign, use ECDHE).
-- If keys are necessary which are outside of the TEE / SE, make sure you obfuscate/encrypt them and only de-obfuscate them during use. Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap.
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use Android Keystore) or Secure Enclave (e.g. use the Keychain and when you sign, use ECDHE).
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. Another approach is obfuscation: obfuscat ...
```
I suggest we do both!

First, we explain the DEK-KEK approach, because it is platform- and implementation-independent. Anyone can do it.

Second, I also added a link to EncryptedSharedPreferences from `androidx.security.crypto` package as an illustration of this approach. I've seen many similar libs for iOS and Android (encrypted shared preferences, encrypted user defaults), but I'd avoid linking community-written libs (because who knows how they have implemented encryption piece, how they store keys, if they provide any expiration/cleanup, etc :) ). 

Linking `androidx.security.crypto` makes sense because it's made by Google, more trust, more support.

I rewrote this piece, please take a look.

vixentael replied:
Random idea: we can also link to this [Cryptographic Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html#encrypting-stored-keys) that also explains key wrapping technique.

vixentael replied:
Also, Amazon calls this technique "envelope encryption": https://docs.aws.amazon.com/wellarchitected/latest/financial-services-industry-lens/use-envelope-encryption-with-customer-master-keys.html

vixentael commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -160,10 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
-- Make sure that keys live in memory for the shortest time possible. Ideally, keys should be read/allocated right before usage and clean up after usage or in case of error. Never store keys in constants.
-- Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap. See [Clean memory of secret data](https://github.com/veorq/cryptocoding#clean-memory-of-secret-data/ "The Cryptocoding Guidelines by @ve ...
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use [Android Keystore](https://developer.android.com/training/articles/keystore.html "Android keystore system")) or [Secure Enclave](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave "Storing Keys in the Secure ...
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. An illustration of this approach is [Encr ...
```
update here

vixentael commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -158,8 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use Android Keystore) or Secure Enclave (e.g. use the Keychain and when you sign, use ECDHE).
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. Another approach is obfuscation: obfuscat ...
+- Make sure that keys live in memory for the shortest time possible. Ideally, keys should be read/allocated right before usage and clean up after usage or in case of error. Never store keys in constants.
+- Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap. See [Clean memory of secret data](https://github.com/veorq/cryptocoding#clean-memory-of-secret-data/ "The Cryptocoding Guidelines by @ve ...
```
Yes, totally makes sense, if we touch zeroing, let's update it fully :)

I didn't find any guaranteed method of zeroing keys for iOS/Android, but I found and linked recommendations (still, no guarantees, but more likely).

I also rewrote wording into less strict, more recommendation/try-your-best style: if we can't controls memory management of operating system, we can't really demand developers to use any certain approach.

Please let me know what do you think

vixentael commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -160,10 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use [Android Keystore](https://developer.android.com/training/articles/keystore.html "Android keystore system")) or [Secure Enclave](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave "Storing Keys in the Secure ...
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. An illustration of this approach is [Encr ...
+- Make sure that keys live in memory for the shortest time possible. Ideally, keys should be read/allocated right before usage and clean up after usage or in case of error. Avoid storing keys in constants.
+- Consider zeroing out and nullify keys after successful cryptographic operation and in case of error. Unfortunately, there are certain limitations of realibly cleaning up secret data in languages with garbage collector (Java) and immutable `String` (Swift, Objective-C, Kotlin). [Java Cryptography Architecture Reference Guide](https://docs.oracle.com/en/java/javase/16/security/java-cryptography-a ...
```
@cpholguera rewritten here

cpholguera commented:
Short update: half review has already happened. Hopefully I can send you some feedback this week. We need to ensure that we properly address these topics, we might have to relocate and merge some info with other parts of the guide. I'll let you know as soon as I have something new. 

btw. thanks for the link to the key wrapping, that's what I was looking for, I couldn't recall the technique myself and I'd like to ensure that we keep using standard wording/naming, etc.

vixentael commented:
thank you @cpholguera! 
I'm open to discussion here and in owasp slack (@vixentael)

cpholguera commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -158,8 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
 
-- make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use Android Keystore) or Secure Enclave (e.g. use the Keychain and when you sign, use ECDHE).
-- If keys are necessary which are outside of the TEE / SE, make sure you obfuscate/encrypt them and only de-obfuscate them during use. Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap.
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use [Android Keystore](https://developer.android.com/training/articles/keystore.html "Android keystore system")) or [Secure Enclave](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave "Storing Keys in the Secure ...
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. An illustration of this approach is [Encr ...
```
```suggestion
- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use [Android Keystore](https://developer.android.com/training/articles/keystore.html "Android keystore system")) or [Secure Enclave](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave "Storing Keys in the Secure Enclave") (e.g. use the Keychain). Refer to the [Android Data Storage](0x05d-Testing-Data-Storage.md#storing-a-key---example) and [iOS Data Storage](0x06d-Testing-Data-Storage.md#the-keychain) chapters for more information.
```

This point is addressing one special case of the previous point for Android, I'd suggest to remove it since the previous point covers it. The current paragraph should be reduced and the specifics moved/merged to their corresp. chapters.

You can use the info from the deleted paragraph to update [Android Data Storage - Storing a key](https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05d-Testing-Data-Storage.md#storing-a-key---example) section. We need a nice overview / clear recommendation  there (e.g. most secure if it never leaves the KeyStore, 2nd most secure if using EncryptedPreferences, etc.). You could take the chance to introduce the official term "Envelope encryption" from [the site you've sent me](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html#encrypting-stored-keys) which refers to [Google](https://cloud.google.com/kms/docs/envelope-encryption?hl=en).

Maybe it could be a nice idea to include another example, since this is also good to know for other kind of apps such as the "React Native"-based ones (and goes in-line with our statement/recommendation here): 
- https://reactnative.dev/docs/security
- https://github.com/emeraldsanto/react-native-encrypted-storage/blob/master/android/src/main/java/com/emeraldsanto/encryptedstorage/RNEncryptedStorageModule.java

Unfortunately I could not find more examples for other frameworks, only this [lib for Flutter](https://github.com/mogol/flutter_secure_storage/blob/develop/android/src/main/java/com/it_nomads/fluttersecurestorage/ciphers/StorageCipher18Implementation.java) , but it has a custom implementation.

cpholguera replied:
I'm realizing now that the location of the mentioned information is not yet consistent. The iOS chapter should have an introductory section regarding iOS Data Storage (reflecting the Android one), instead of inside the test case. 

  > This refactoring should be eventually made. If you want please go ahead and move/consolidate that info (as part of a new PR).

cpholguera commented on Document/0x04g-Testing-Cryptography.md:
```diff
@@ -158,8 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
-- If keys are necessary which are outside of the TEE / SE, make sure you obfuscate/encrypt them and only de-obfuscate them during use. Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap.
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use [Android Keystore](https://developer.android.com/training/articles/keystore.html "Android keystore system")) or [Secure Enclave](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave "Storing Keys in the Secure ...
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. An illustration of this approach is [Encr ...
+- Make sure that keys live in memory for the shortest time possible. Ideally, keys should be read/allocated right before usage and clean up after usage or in case of error. Avoid storing keys in constants.
+- Consider zeroing out and nullify keys after successful cryptographic operation and in case of error. Unfortunately, there are certain limitations of realibly cleaning up secret data in languages with garbage collector (Java) and immutable `String` (Swift, Objective-C, Kotlin). [Java Cryptography Architecture Reference Guide](https://docs.oracle.com/en/java/javase/16/security/java-cryptography-a ...
```
```suggestion
- Make sure that keys live in memory for the shortest time possible and consider zeroing out and nullifying keys after successful cryptographic operations, or in case of error. For more detailed information refer to sections [Testing Memory for Sensitive Data](0x05d-Testing-Data-Storage.md#checking-memory-for-sensitive-data-mstg-storage-10) and [Checking Memory for Sensitive Data](0x05d-Testing-Data-Storage.md#checking-memory-for-sensitive-data-mstg-storage-10) for Android and iOS respectively.
```

To keep it simple and general (since we're in the general guide here), I suggest to reduce these two points to the suggestion and use the _removed_ information to update these sections:

- https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05d-Testing-Data-Storage.md#checking-memory-for-sensitive-data-mstg-storage-10
- https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06d-Testing-Data-Storage.md#testing-memory-for-sensitive-data-mstg-storage-10

Also consider renaming the section from "Checking Memory for Sensitive Data" to "Testing Memory for Sensitive Data" for consistency.

cpholguera reviewed (changes requested):
Hi @vixentael, the review took some time since you're touching very interesting points here. I was thinking that this give us the chance to improve the guide even more, so I wanted to carefully craft the suggestions. I hope you like them and can help us out with a little refactoring / update of some sections. We should keep this chapter "general" and move the specifics to the Android/iOS chapters.

I'd understand if something is not completely clear, please feel free to discuss here or in Slack :)

Thanks a lot again for taking care of this!

vixentael commented:
@cpholguera thank you for the review! I totally support the idea of having general details in general guide and moving specific details into iOS/Android guides.

---
Things I've done:

**0x04g-Testing-Cryptography**

1. Reduced details in key management chapter. I left only cross-platform things and introduced *envelope encryption* technique (because it's 100% cross platform :) ).

**0x05d-Testing-Data-Storage** (Android)

1. Updated Android "Storing a cryptographic key: techniques" chapter, where I added more details into "Storing a Key - from most secure to least secure" list, updated "key clean up" info, described envelope encryption, and linked with Android EncryptedSharedPreferences library.

2. Renamed "Checking Memory for Sensitive Data" to "Testing Memory for Sensitive Data", and fixed all mentions/links of the previous name. 

**0x06d-Testing-Data-Storage** (iOS)

1. Added short details about different key storage ways, and some info about sensitive data clean up.

---
Things I haven't done:

1. I haven't used react native libs examples that you've mentioned. Because we decided to have general chapter more general, it doesn't have links to platform-specific libraries now. And AFAIK, MSTG is targeted on native apps mainly. Where do you see RN libraries fit? 
2. As iOS testing guide doesn't have a Theory Overview chapter (yet), I decided to add just small pieces about key management. I am up to making a separate PR with theory chapter that describes preferred key management ways for iOS similar to what we have for Android. But before me doing that, let's agree on this PR first :)

vixentael commented:
@cpholguera I mean no rush, but what do you think about the latest update? I'd like to get it sorted and to move to updating iOS crypto part next.

cpholguera commented:
Hi, we've been very busy with the MASVS release and now on leave until June. We'll go back to this one then. Thanks for the patience!

vixentael commented:
@cpholguera no mean to rush, but Apple has introduced [AES.KeyWrap API](https://developer.apple.com/documentation/cryptokit/aes/keywrap) on the latest WWDC :) (No docs now, but it's similar to https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey)

It means that after merging this PR, I'll have even more things-to-write about in a new chapter in iOS Testing guide :)

cpholguera commented:
Sounds very nice, is there a video for that? 
Please feel free to open an issue 😉

Thanks for your patience, we hope we can get back to the PRs soon. There's a lot going on right now. We'll announce something soon, I'm sure you'll like it :)

cpholguera commented:
Hi @vixentael, just to let you know: we did not forget about this PR. As you know we're currently performing a big refactor of the MASVS and your content in this PR is helping us a lot currently since we're taking a look at V2 specifically :)

Thanks for your patience!

vixentael commented:
hi @cpholguera thank you for the update! I keep an eye on discussions in MASVS, but I'm also available in OWASP slack, so if any questions – i'd be happy to help in shaping crypto requirements.

cpholguera reviewed (changes requested):
Thanks a lot @vixentael and sorry about the long wait. I hope you're doing well.

I'm proposing changes I got from a review I started doing some time ago. Nothing for you to change for now. @TheDauntless will help with the review by first reviewing my review (if that makes sense :) ).

However, feel free to comment if you want to.

TheDauntless commented on Document/0x05d-Testing-Data-Storage.md:
```diff
@@ -391,22 +391,45 @@ Storing a Key - from most secure to least secure:
 A less secure way of storing encryption keys, is in the SharedPreferences of Android. When [SharedPreferences](https://developer.android.com/reference/android/content/SharedPreferences.html "Android SharedPreference API") are used, the file is only readable by the application that created it. However, on rooted devices any other application with root access can simply read the SharedPreference fi ...
 
-The last three options are to use hardcoded encryption keys in the source code, having a predictable key derivation function based on stable attributes, and storing generated keys in public places like `/sdcard/`. Obviously, hardcoded encryption keys are not the way to go. This means every instance of the application uses the same encryption key. An attacker needs only to do the work once, to ext ...
+The last three options are to use hardcoded encryption keys in the source code, having a predictable obfuscation function or key derivation function based on stable attributes, and storing generated keys in public places like `/sdcard/`. Obviously, hardcoded encryption keys are not the way to go. This means every instance of the application uses the same encryption key. An attacker needs only to  ...
```
```suggestion
The last three options are to use hardcoded encryption keys in the source code, having a predictable obfuscation function or key derivation function based on stable attributes, and storing generated keys in public places like `/sdcard/`. Hardcoded encryption keys are an issue since this means every instance of the application uses the same encryption key. An attacker can reverse-engineer a local copy of the application in order to extract the cryptographic key, and use that key to decrypt any data which was encrypted by the application on any device.
```

TheDauntless reviewed (changes requested):
Some additional changes.

cpholguera reviewed (approved):
@vixentael this is awesome content, we've applied some fixes and will merge. We hope to have your support for the upcoming MSTG refactoring!
